### ISSUE TITLE
Error: Failed assertion: mismatched composition of two changesets

### ISSUE THREAD

**Issue by JohnMcLear** [ether/etherpad-lite#2455]
Had this on beta.etherpad.org, brought server down :)

```
[2015-01-12 20:52:40.544] [ERROR] console - Error: Failed assertion: mismatched composition of two changesets
    at Object.error (/home/jose/etherpad-lite/src/static/js/Changeset.js:39:11)
    at Object.assert (/home/jose/etherpad-lite/src/static/js/Changeset.js:53:13)
    at Object.exports.compose (/home/jose/etherpad-lite/src/static/js/Changeset.js:1319:11)
    at /home/jose/etherpad-lite/src/node/handler/PadMessageHandler.js:1636:31
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:610:21
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:249:17
    at iterate (/home/jose/etherpad-lite/src/node_modules/async/lib/async.js:149:13)
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:160:25
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:251:21
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:615:34
```

**Comment by JohnMcLear** [comment]
@marcelklehr @webzwo0i @dmitryuv any ideas on this one?

**Comment by dmitryuv** [comment]
put more debug info to figure out bad changeset, and/or wrap this part into try/catch

**Comment by JohnMcLear** [comment]
@dmitryuv any tips for where I might get value from adding debug or which part/lines to wrap in a try catch?

**Comment by dmitryuv** [comment]
at /home/jose/etherpad-lite/src/node/handler/PadMessageHandler.js:1636:31

the composition loop around this row can throw on bad changeset, this needs to be handled gracefully.

**Comment by JohnMcLear** [comment]
Yea that's where I put it, let's see if we get an exception within the next 24 hours.

**Comment by JohnMcLear** [comment]
Does this help?

```
 cs2:Z:2km>1|h=247=gd*0|1+1$
pool:[object Object]unpacked1:[object Object]unpacked2:[object Object]
    at Object.error (/home/jose/etherpad-lite/src/static/js/Changeset.js:39:11)
    at Object.assert (/home/jose/etherpad-lite/src/static/js/Changeset.js:54:13)
    at Object.exports.compose (/home/jose/etherpad-lite/src/static/js/Changeset.js:1320:11)
    at /home/jose/etherpad-lite/src/node/handler/PadMessageHandler.js:1635:31
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:610:21
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:249:17
    at iterate (/home/jose/etherpad-lite/src/node_modules/async/lib/async.js:149:13)
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:160:25
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:251:21
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:615:34
 [31m[2015-02-03 17:13:56.825] [ERROR] console -  [39mERROR { oldLen: 3476,
  newLen: 3428,
  ops: '|32=2n5|2-r-m*a|1+1',
  charBank: '
' }
[2015-02-03 17:13:56.825] [ERROR] console - ERROR { oldLen: 3476,
  newLen: 3428,
  ops: '|32=2n5|2-r-m*a|1+1',
  charBank: '
' }
 [31m[2015-02-03 17:13:56.826] [ERROR] console -  [39mError: Failed assertion: [object Object][object Object]mismatched composition of two changesets..... cs1:Z:2ok<1c|32=2n5|2-r-m*a|1+1$
cs2:Z:2mc>6|j=2km=13*5+6$ with pool:[object Object]unpacked1:[object Object]unpacked2:[object Object]
    at Object.error (/home/jose/etherpad-lite/src/static/js/Changeset.js:39:11)
    at Object.assert (/home/jose/etherpad-lite/src/static/js/Changeset.js:54:13)
    at Object.exports.compose (/home/jose/etherpad-lite/src/static/js/Changeset.js:1320:11)
    at /home/jose/etherpad-lite/src/node/handler/PadMessageHandler.js:1635:31
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:610:21
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:249:17
    at iterate (/home/jose/etherpad-lite/src/node_modules/async/lib/async.js:149:13)
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:160:25
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:251:21
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:615:34
[2015-02-03 17:13:56.826] [ERROR] console - Error: Failed assertion: [object Object][object Object]mismatched composition of two changesets..... cs1:Z:2ok<1c|32=2n5|2-r-m*a|1+1$
cs2:Z:2mc>6|j=2km=13*5+6$ with pool:[object Object]unpacked1:[object Object]unpacked2:[object Object]
    at Object.error (/home/jose/etherpad-lite/src/static/js/Changeset.js:39:11)
    at Object.assert (/home/jose/etherpad-lite/src/static/js/Changeset.js:54:13)
    at Object.exports.compose (/home/jose/etherpad-lite/src/static/js/Changeset.js:1320:11)
    at /home/jose/etherpad-lite/src/node/handler/PadMessageHandler.js:1635:31
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:610:21
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:249:17
    at iterate (/home/jose/etherpad-lite/src/node_modules/async/lib/async.js:149:13)
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:160:25
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:251:21
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:615:34
```

**Comment by dmitryuv** [comment]
there should be 2 changesets for composition operation :) The second one is unpacked? Then lengths do not match, they can't be composed.

**Comment by JohnMcLear** [comment]
changeset 1 =  Z:2ok<1c|32=2n5|2-r-m*a|1+1$ 
changeset 2 = Z:2mc>6|j=2km=13*5+6$

Does that help?

**Comment by webzwo0i** [comment]
not sure if I can come back to this soon...

in decimal
3476<48|110=3425|2-27-22*a|1+1$ (oldlen:3476, newlen: 3428, 3425 chars + 27 + 22 + 1 = 3475 )
3396>6|19=3334=39*5+6$ (oldlen:3396,newlen:3402, 3334 chars + 39 + 6 = 3379 )

the first is off by one, but I think I have seen this behaviour due to the "\n" that is appended (I don't know why exactly) sometimes at the end. this cs basically imo says that somebody entered newline at this end of the pad.
but the second seems as if something is wrong... its off by 20! its an attribute change-only cs (attribute *5 is applied to position 3373 until 3379) but I wonder why it says ">6" (6 chars are added).

**Comment by webzwo0i** [comment]
could you possibly share the pad content or look by yourself if something special is at those positions? (e.g. a link are something)
to get the pad with mysql:
mysql> select * from `store` where `value` like '%Z::2ok<1c|32=2n5|2-r-m*a|1+1%'; should give you pad&rev, then http://installation/p/the-pad-url/the-rev-found/export/html from this rev and the rev before and after?

**Comment by JohnMcLear** [comment]
That's a great idea..  I will get onto this now.

**Comment by JohnMcLear** [comment]
That select didn't show anything, perhaps it wasn't written to the database as it failed?

I did however notice this...

```
mysql> SHOW FULL COLUMNS FROM store;
+-------+--------------+-------------------+------+-----+---------+-------+---------------------------------+---------+
| Field | Type         | Collation         | Null | Key | Default | Extra | Privileges                      | Comment |
+-------+--------------+-------------------+------+-----+---------+-------+---------------------------------+---------+
| key   | varchar(100) | utf8_general_ci   | NO   | PRI |         |       | select,insert,update,references |         |
| value | longtext     | latin1_swedish_ci | NO   |     | NULL    |       | select,insert,update,references |         |
+-------+--------------+-------------------+------+-----+---------+-------+---------------------------------+---------+
2 rows in set (0.01 sec)

mysql> alter table store convert to character set utf8 
    -> ;
Query OK, 632400 rows affected (14.14 sec)
Records: 632400  Duplicates: 0  Warnings: 0

mysql> SHOW FULL COLUMNS FROM store;
+-------+--------------+-----------------+------+-----+---------+-------+---------------------------------+---------+
| Field | Type         | Collation       | Null | Key | Default | Extra | Privileges                      | Comment |
+-------+--------------+-----------------+------+-----+---------+-------+---------------------------------+---------+
| key   | varchar(100) | utf8_general_ci | NO   | PRI |         |       | select,insert,update,references |         |
| value | longtext     | utf8_general_ci | NO   |     | NULL    |       | select,insert,update,references |         |
+-------+--------------+-----------------+------+-----+---------+-------+---------------------------------+---------+
2 rows in set (0.00 sec)

mysql> 
```

Note the encoding type..

**Comment by JohnMcLear** [comment]
@webzwo0i any idea how to get the padId knowing that the select dosen't work?

**Comment by webzwo0i** [comment]
strange code is:
      changeset = changesets[startNum];
      var pool = pad.apool();

      for(var r=startNum+1;r<endNum;r++)
      { 
        var cs = changesets[r];
        changeset = Changeset.compose(changeset, cs, pool);
      } 
changesets was filled before with rev->changesets mapping so both values should be there o.O
can you check if non of the above changesets is in the database?
also: it is called via PadMessageHandler.js:1636 (composePadChangesets). You could grep all you're plugins to see which actually use this function.
In the core node I only see one usage, namely in getChangesetInfo via handleChangesetRequest which gets called on CHANGESET_REQ messages.
Those are only used in timeslider-code right?

**Comment by webzwo0i** [comment]
yeah only in broadcast.js which is used in timeslider (and maybe in plugins?). If this is right, then we could write a loop that tries to timeslide all pads that where active at about 2015-02-03 17:13:56
so you should get the pads with timestamps between 1422983630 (second 50) and 1422983639 (second 59). could you grep for this? SELECT * from `store` where `value` like '%1422983630%';

**Comment by webzwo0i** [comment]
in the last message the select was wrong it should be something like:
SELECT * from store where value like '%142298363%'; or even SELECT * from store where value like '%14229836%'; maybe you have luck and not to many pads where active at this time.

**Comment by JohnMcLear** [comment]
```
mysql> SELECT * from store where value like '%14229836%';
+---------------------------------+-------------------------------------------------------------------------------------------| key                             | value                                                                                                       |
+---------------------------------+-------------------------------------------------------------------------------------------
| globalAuthor:a.t1UvAkTn7PeXHlSJ | {"colorId":"#c7ffff","name":"The Dude","timestamp":1422983653274,"padIDs":{"Oe34":1}}                       |
| globalAuthor:a.Yh2o5JlEknI3EQXJ | {"colorId":30,"name":null,"timestamp":1422983656630}                                                        |
| globalAuthor:a.i5PJ6uwmgebl49fJ | {"colorId":"#c1d9a9","name":"Enshael","timestamp":1422983655166,"padIDs":{"w8oKkU3veX":1}}                  |
| globalAuthor:a.o9oTHhtrsJ7C8Oum | {"colorId":"#c7ff8f","name":"Piero","timestamp":1422983652604,"padIDs":{"Andrea":1}}                        |
| globalAuthor:a.83R4Mq0IeoTVH79n | {"colorId":"#a979d9","name":"","timestamp":1422983658791}                                                   |
| globalAuthor:a.RG3JjZgfWJtlOZ1I | {"colorId":"#e3c7ff","name":"","timestamp":1422983657794,"padIDs":{"test":1}}                               |
| pad:Andrea:revs:1560            | {"changeset":"Z:1sd>0|7=96*a|9=ji*a=3h$","meta":{"author":"a.OlKoCpEnf9TTRFNY","timestamp":1422983611734}}  |
| pad:Andrea:revs:1561            | {"changeset":"Z:1sd>0|i=w7*a|3=2h*a=27$","meta":{"author":"a.OlKoCpEnf9TTRFNY","timestamp":1422983618266}}  |
| pad:Andrea:revs:1562            | {"changeset":"Z:1sd>0|n=13o*a|1=23*a=25$","meta":{"author":"a.OlKoCpEnf9TTRFNY","timestamp":1422983623826}} |
| pad:Andrea:revs:1563            | {"changeset":"Z:1sd>0|u=1bx*a|4=8e*a=22$","meta":{"author":"a.OlKoCpEnf9TTRFNY","timestamp":1422983628596}} |
| pad:Andrea:revs:1564            | {"changeset":"Z:1sd>0|13=1ny*a|2=1o*a=t$","meta":{"author":"a.OlKoCpEnf9TTRFNY","timestamp":1422983636077}} |
```

**Comment by webzwo0i** [comment]
I probably triggered it right now, is there a stacktrace?

**Comment by webzwo0i** [comment]
could you send me SELECT * from `store` where `key` like '%pad:test:revs%'; via mail?

**Comment by JohnMcLear** [comment]
last one was 2 hours ago:

```
' }
[2015-02-04 08:42:45.477] [ERROR] console - ERROR { oldLen: 689,
  newLen: 693,
  ops: '|4=3b*bf|4+4*f*g=1|1=7*1b=1|1=l*1c=1',
  charBank: '



' }
 [31m[2015-02-04 08:42:45.478] [ERROR] console -  [39mError: Failed assertion: [object Object][object Object]mismatched composition of two changesets..... cs1:Z:j5>4|4=3b*bf|4+4*f*g=1|1=7*1b=1|1=l*1c=1$



cs2:Z:hl<1|e=65=17-1$pool:[object Object]unpacked1:[object Object]unpacked2:[object Object]
    at Object.error (/home/jose/etherpad-lite/src/static/js/Changeset.js:39:11)
    at Object.assert (/home/jose/etherpad-lite/src/static/js/Changeset.js:54:13)
    at Object.exports.compose (/home/jose/etherpad-lite/src/static/js/Changeset.js:1320:11)
    at /home/jose/etherpad-lite/src/node/handler/PadMessageHandler.js:1635:31
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:610:21
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:249:17
    at iterate (/home/jose/etherpad-lite/src/node_modules/async/lib/async.js:149:13)
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:160:25
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:251:21
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:615:34
[2015-02-04 08:42:45.478] [ERROR] console - Error: Failed assertion: [object Object][object Object]mismatched composition of two changesets..... cs1:Z:j5>4|4=3b*bf|4+4*f*g=1|1=7*1b=1|1=l*1c=1$



cs2:Z:hl<1|e=65=17-1$pool:[object Object]unpacked1:[object Object]unpacked2:[object Object]
    at Object.error (/home/jose/etherpad-lite/src/static/js/Changeset.js:39:11)
    at Object.assert (/home/jose/etherpad-lite/src/static/js/Changeset.js:54:13)
    at Object.exports.compose (/home/jose/etherpad-lite/src/static/js/Changeset.js:1320:11)
    at /home/jose/etherpad-lite/src/node/handler/PadMessageHandler.js:1635:31
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:610:21
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:249:17
    at iterate (/home/jose/etherpad-lite/src/node_modules/async/lib/async.js:149:13)
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:160:25
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:251:21
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:615:34
```

there was a restart at 2am (9 hours ago) but it failed to send me a stack trace

@webzwo0i the pad:test:revs sql output of ``SELECT * from store where `key` like "%pad:test:revs%";`` is 5673 rows..  Are you sure you want that much or do you want to use a different padId?

**Comment by webzwo0i** [comment]
yeah 9 hours ago was probably me and the revs would help I think:
mysql -u username -p -e 'select * from `store` where `key` like "pad:test:revs%";' your-database|gzip >test.sql should make it smaller
SELECT * from `store` where `value` like '%142303936%'; should show the active pads/user sessions during the crash before 9am

**Comment by JohnMcLear** [comment]
Email sent.

**Comment by webzwo0i** [comment]
the changesets are definitely broken... some contain to many backslashes. in case of "\\n" this might lead to a newline in the cs-metadata when in reality it will be just a normal character. this could imo cause the character mismatches crashes
same thing happens for '\\"'. this will essentially unescape the second backslash so that there is a double-quote left. this will crash all code that things this is valid JSON

**Comment by webzwo0i** [comment]
you could try in the database:
mysql> select * from `store` where `value` like '\\\\"' limit 1; resp
mysql> select * from `store` where `value` like '\\"' limit 1; resp
mysql> select * from `store` where `value` like '\\n' limit 1;

to see when they first appeared. Maybe the timestamps could help us to find out if its a commit in epl or in a plugin...

**Comment by JohnMcLear** [comment]
None of the plugins on beta.etherpad.org write directly to the database FWIW

**Comment by JohnMcLear** [comment]
Just use three back ticks to represent a code block, I keep modifying your comments using these ;)  Yuo can just click the pencil to see how I do it!

**Comment by JohnMcLear** [comment]
Note that none of these are in the atext:
```
mysql> select * from `store` where `value` like '%\\n%' limit 1;
+-----------------------------------------+-------------------------------------------------------------------------------+
| key                                     | value                                                                         |
+-----------------------------------------+-------------------------------------------------------------------------------+
| sessionstorage:5jXsdrUPQAGcgNrsuc6JtH1Z | {"cookie":{"path":"/","_expires":null,"originalMaxAge":null,"httpOnly":true}} |
+-----------------------------------------+-------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> select * from `store` where `value` like '%\\%"' limit 1; 
^[[A^[[A^[[AEmpty set (1.00 sec)

mysql> select * from `store` where `value` like '%\\\\%"' limit 1;
+-----------------------+----------------------------------+
| key                   | value                            |
+-----------------------+----------------------------------+
| custom_style_css_test | ".test{\n  margin-left:20px;\n}" |
+-----------------------+----------------------------------+
1 row in set (0.10 sec)
```

**Comment by webzwo0i** [comment]
strange. Must think about this again...
Any crash happened during the last 45 minutes? I tested the timeslider and export of the pads you sent me, thats why I am curious.

**Comment by JohnMcLear** [comment]
22 minutes ago

```
' }
[2015-02-04 15:40:27.046] [ERROR] console - ERROR { oldLen: 689,
  newLen: 693,
  ops: '|4=3b*bf|4+4*f*g=1|1=7*1b=1|1=l*1c=1',
  charBank: '



' }
 [31m[2015-02-04 15:40:27.047] [ERROR] console -  [39mError: Failed assertion: [object Object][object Object]mismatched composition of two changesets..... cs1:Z:j5>4|4=3b*bf|4+4*f*g=1|1=7*1b=1|1=l*1c=1$



cs2:Z:hl<1|e=65=17-1$pool:[object Object]unpacked1:[object Object]unpacked2:[object Object]
    at Object.error (/home/jose/etherpad-lite/src/static/js/Changeset.js:39:11)
    at Object.assert (/home/jose/etherpad-lite/src/static/js/Changeset.js:54:13)
    at Object.exports.compose (/home/jose/etherpad-lite/src/static/js/Changeset.js:1320:11)
    at /home/jose/etherpad-lite/src/node/handler/PadMessageHandler.js:1635:31
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:610:21
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:249:17
    at iterate (/home/jose/etherpad-lite/src/node_modules/async/lib/async.js:149:13)
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:160:25
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:251:21
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:615:34
[2015-02-04 15:40:27.047] [ERROR] console - Error: Failed assertion: [object Object][object Object]mismatched composition of two changesets..... cs1:Z:j5>4|4=3b*bf|4+4*f*g=1|1=7*1b=1|1=l*1c=1$



cs2:Z:hl<1|e=65=17-1$pool:[object Object]unpacked1:[object Object]unpacked2:[object Object]
    at Object.error (/home/jose/etherpad-lite/src/static/js/Changeset.js:39:11)
    at Object.assert (/home/jose/etherpad-lite/src/static/js/Changeset.js:54:13)
    at Object.exports.compose (/home/jose/etherpad-lite/src/static/js/Changeset.js:1320:11)
    at /home/jose/etherpad-lite/src/node/handler/PadMessageHandler.js:1635:31
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:610:21
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:249:17
    at iterate (/home/jose/etherpad-lite/src/node_modules/async/lib/async.js:149:13)
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:160:25
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:251:21
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:615:34
```

If it helps I can give you access to the box?

**Comment by dmitryuv** [comment]
Guys, check the database encoding and encoding between client and server, everything should be unicode

**Comment by JohnMcLear** [comment]
@dmitryuv yea see my previous comments..  You are right though..  It could still be that as a cause of the issue..

**Comment by webzwo0i** [comment]
@dmitryuv  I don't understand the details well enough, but  shouldn't there be some non-ascii characters in the specific changesets that cause the crashes? I did not check all of them, but at least some have ascii-only characters and still wrong backslash-escaping.

I noticed that there where some mysql-related changes in ueberDB last October/November, I think I am gonna test without them

**Comment by dmitryuv** [comment]
There can be some Unicode stuff like space or smthng when copying from another source, for example.

**Comment by webzwo0i** [comment]
I tried to rebuild the latin1 situation on another mysql installation but wasn't able to reproduce. Probably gonna give up on this one :(

**Comment by JohnMcLear** [comment]
@webzwo0i I think both myself and @dmitryuv have confused you here.  The issue isn't due to database encoding.

Even with beta.etherpad.org having the correct encoding we still get the error.

You said you were able to replicate a crash the other night, how did you do that?

Latest crash was 2 hours ago
```
 cs2:Z:hl<1|e=65=17-1$pool:[object Object]unpacked1:[object Object]unpacked2:[object Object]
    at Object.error (/home/jose/etherpad-lite/src/static/js/Changeset.js:39:11)
    at Object.assert (/home/jose/etherpad-lite/src/static/js/Changeset.js:54:13)
    at Object.exports.compose (/home/jose/etherpad-lite/src/static/js/Changeset.js:1320:11)
    at /home/jose/etherpad-lite/src/node/handler/PadMessageHandler.js:1635:31
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:610:21
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:249:17
    at iterate (/home/jose/etherpad-lite/src/node_modules/async/lib/async.js:149:13)
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:160:25
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:251:21
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:615:34
 [31m[2015-02-07 11:07:27.687] [ERROR] console -  [39mERROR { oldLen: 2294,
  newLen: 2310,
  ops: '|q=1pl=23*cg+g',
  charBank: ' ainsi que je tr' }
[2015-02-07 11:07:27.687] [ERROR] console - ERROR { oldLen: 2294,
  newLen: 2310,
  ops: '|q=1pl=23*cg+g',
  charBank: ' ainsi que je tr' }
 [31m[2015-02-07 11:07:27.688] [ERROR] console -  [39mError: Failed assertion: [object Object][object Object]mismatched composition of two changesets..... cs1:Z:1rq>g|q=1pl=23*cg+g$ ainsi que je trcs2:Z:1sr>1|6=9j=1n*ch+1$gpool:[object Object]unpacked1:[object Object]unpacked2:[object Object]
    at Object.error (/home/jose/etherpad-lite/src/static/js/Changeset.js:39:11)
    at Object.assert (/home/jose/etherpad-lite/src/static/js/Changeset.js:54:13)
    at Object.exports.compose (/home/jose/etherpad-lite/src/static/js/Changeset.js:1320:11)
    at /home/jose/etherpad-lite/src/node/handler/PadMessageHandler.js:1635:31
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:610:21
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:249:17
    at iterate (/home/jose/etherpad-lite/src/node_modules/async/lib/async.js:149:13)
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:160:25
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:251:21
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:615:34
[2015-02-07 11:07:27.688] [ERROR] console - Error: Failed assertion: [object Object][object Object]mismatched composition of two changesets..... cs1:Z:1rq>g|q=1pl=23*cg+g$ ainsi que je trcs2:Z:1sr>1|6=9j=1n*ch+1$gpool:[object Object]unpacked1:[object Object]unpacked2:[object Object]
    at Object.error (/home/jose/etherpad-lite/src/static/js/Changeset.js:39:11)
    at Object.assert (/home/jose/etherpad-lite/src/static/js/Changeset.js:54:13)
    at Object.exports.compose (/home/jose/etherpad-lite/src/static/js/Changeset.js:1320:11)
    at /home/jose/etherpad-lite/src/node/handler/PadMessageHandler.js:1635:31
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:610:21
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:249:17
    at iterate (/home/jose/etherpad-lite/src/node_modules/async/lib/async.js:149:13)
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:160:25
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:251:21
    at /home/jose/etherpad-lite/src/node_modules/async/lib/async.js:615:34
```

**Comment by webzwo0i** [comment]
TLDR: latin1 is really bad here, but it's not the root cause of this issue I guess.

I think you were right with the utf8 stuff, but the problem is not easily solved though. mysql converted utf8 to latin1 at the column level. This is a one-way thing and cannot automatically be reversed, because the data is permanently destroyed in the database (if other people experience the same they might have luck if they run epl with debug-logging because it logs all the cs).
What happens is that valid chars (in utf8 encoding) end up as single or more ? in the db. The details exceed my knowledge, I just noticed that for chars with an utf8 encoding between 1-3 bytes a single ? ends up in db, for chars with 4 byte utf8 encoding 4 ? are saved.
So the chars are gone. But what makes it even harder to fix is that we have the length-fields in our cs operators that depend on the sanity of the text (think of the position of newlines or in general the text length). I think different kinds of corruption could happen here.

As said above I cannot reproduce with a fresh installation, but I can trigger the bug on beta in two different ways on some pads. via timeslider and via export. Both access the length properties of changesets at some time and either compare it to the length of text or to the previous changeset's length+the amount that changeset actually changes.

https://github.com/webzwo0i/ep-debug-helper/blob/master/check-length.rb iterates over all changesets and complains, if for two cs: (cs1 +/- chars changed) != cs2.length
With this small check I found 13 pads and 20 changesets that are stored but have a wrong "oldlength"; manual inspection does not help, they seem to look okay besides that. The only thing I noticed is: mysql is configured to be case-insensitive and in every occasion where the wrong cs has been stored, multiple users were involved that where connected via different pad urls. It is not deterministic (at least I think so), so sometimes user a was accessing the pad via URL and another via Url.
Interesting enough in every case its like this:
pad:somepad:rev:X
pad:somepad:rev:X+1
pad:Somepad:rev:X+2 <--bad cs (so *always* the bad cs is one that is not case-sensitive equal to the previous)
"Not deterministic" means, it does not matter if lower case follows upper case or the other way around. Also during normal pad operations its possible that people access the pad via all kinds of possible combinations of how to write the pad considering case-insensitivity, without getting broken changesets.

I am going to extend the script to include the actual atexts and cs-text additions and see if this helps. Also during debugging I found some other minor issues and at least one other reproducible crash condition for which I will create separate issues. Here are some of the changesets I was talking about. I obfuscated the pad names (but keep how they changed in case) and the author ids. Notice the last pad, where author id is empty. I don't know if this is wanted behaviour but it means, that some editor added the newline and not a real user. It happens e.g. when you see delete everything, and enter a new letter. Then automagically a newline is added at the end, with an empty author attribute:
```
pad:Abc:revs:158        {"changeset":"Z:8>1|1=1=5*0|1+1$\\n","meta":{"author":"a.1","timestamp":1422962923412}}
pad:AbC:revs:159        {"changeset":"Z:9<1|1=1=4-1$","meta":{"author":"a.2","timestamp":1422962936994}}
pad:AbC:revs:160        {"changeset":"Z:8<2|1=1=2-2$","meta":{"author":"a.2","timestamp":1422962937509}}
pad:AbC:revs:161        {"changeset":"Z:6<1|1=1=1-1$","meta":{"author":"a.2","timestamp":1422962938008}}
pad:Abc:revs:162        {"changeset":"Z:6>1|3=3=1*0+1$o","meta":{"author":"a.1","timestamp":1422962934587}}

pad:Abc:revs:3   {"changeset":"Z:4n<4m|1-4m$","meta":{"author":"a.1","timestamp":1422962087845}}
pad:Abc:revs:1   {"changeset":"Z:e7<9l=k|8-9l$","meta":{"author":"a.1","timestamp":1422962085012}}
pad:Abc:revs:2   {"changeset":"Z:4m>1=4l|1+1$\\n","meta":{"author":"","timestamp":1422962085012}}
pad:abc:revs:4   {"changeset":"Z:2>1|1=1*0+1$l","meta":{"author":"a.2","timestamp":1422962134113}}

pad:abc:revs:3       {"changeset":"Z:22<1f|1=l|2-1g*0|1+1$\\n","meta":{"author":"a.1","timestamp":1422962816845}}
pad:abc:revs:4       {"changeset":"Z:n<k|1-l*0|1+1$\\n","meta":{"author":"a.1","timestamp":1422962817793}}
pad:abc:revs:1       {"changeset":"Z:e7<c6|1=l|7-9k-2n*0|1+1$\\n","meta":{"author":"a.1","timestamp":1422962813880}}
pad:abc:revs:2       {"changeset":"Z:21>1|2=m=1e|1+1$\\n","meta":{"author":"","timestamp":1422962813880}}
pad:Abc:revs:5       {"changeset":"Z:l>1=k|1+1$\\n","meta":{"author":"","timestamp":1422962782276}}
```
I could also share all of them or the actual pad urls...
One last thing: I just noticed there exist binlogs and I will try to inspect them to find texts after one of those broken changesets...

**Comment by webzwo0i** [comment]
shit...binlogs were not active

**Comment by JohnMcLear** [comment]
Hrm RE mysql not being case sensitive, isn't that a huge problem then?  We would need to .toLowerCase() all padIds?

**Comment by webzwo0i** [comment]
Probably the best idea. Users can still write what they want, as soon as toLowerCase() is equal to another pad they see the version that was first.
If we don't do this: globalAuthor has all the different spellings saved to database. Their might be other, more serios problems, if we somewhere use the pad name as key in javascript objects...
Another thing I noticed, iirc it was like this:
Author1 creates a pad with spelling A (this results in pad:spellingA:revs:0 saved to db)
Author2 does some changes with spelling B (results in pad:SpellingB:revs:X saved to db and and a version of the atext saved as pad:SpellingB - whenever a new version arrives this key changes spelling to the version of the cs)
Author3 uses SpellingA and an author attrib value like "0" (that actually is Author2 because when creating a pad the author id is not saved and Author1 did make any changes and it is also not catched when we check if a user tries to submit something as another author. I think somehow transparently the attrib is changed to the right number "1") -> epl tries to apply the changeset to revision 0!! So at least there must be some code that depends on the spelling.
The changeset is wrong (because it should apply to 3) and is discarded, but I think those kind of problems would arise if we don't unify the spellings in the first place.

However, I wonder how this will affect active sessions or something else...
I did a short grep for "pad" in src/node and probably we could manually check if we somewhere depend on spelling

**Comment by JohnMcLear** [comment]
@webzwo0i are you sure mysql isn't case sensitive when setup using the dosc from the wiki?
https://github.com/ether/etherpad-lite/wiki/How-to-use-Etherpad-Lite-with-MySQL

http://stackoverflow.com/questions/15686944/mysql-case-sensitivity-for-primary-key & http://stackoverflow.com/questions/18737805/mysql-case-sensitive-in-utf8-general-ci

It looks like how we configure the database we shouldn't have that problem...

**Comment by webzwo0i** [comment]
thanks that you figured this out! I would have started doing this code revamping for nothing ;)
I recognized that beta.etherpad.org was configured with "ci" and not "utf8_bin" and because I wanted to reproduce using the exact some conditions I created my db not according to the tutorial...
OK now....
so I will try the "is-this-changeset-ok"-stuff and follow the text in any revision. If this shows up nothing I would vote for "something strange happened due to bad encoding". We would still have to clean out all those broken pads or fix the timeslider and export to not crash if a changeset in the database is wrong and see if this error ever comes back again...

**Comment by dmitryuv** [comment]
> I don't know if this is wanted behaviour but it means, that some editor added the newline and not a real user. It happens e.g. when you see delete everything, and enter a new letter. Then automagically a newline is added at the end

etherpad likes newlines at the end of the document, so if there isn't any, it adds one by itself, with empty author.

**Comment by JohnMcLear** [comment]
This appears to be resolved now, this is basically part of #2516 now so will be closed =)
